### PR TITLE
Dialogue: new approach to edit timestamp

### PR DESCRIPTION
**Pull request by retrofox**
Fixes https://github.com/Automattic/jetpack/issues/18675

#### Changes proposed in this Pull Request:

* Dialogue: new approach to edit timestamp

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Create/edit a post
* Add a podcast block with a valid RSS feed. You can use this one: `https://distributed.blog/2020/07/24/lara-hogan-management-secrets/`
* Add a Conversation/Dialogue blocks composition
* you should be able to see the media player control in the Dialogue toolbar, as well as the `Add timestamp` link in the Dialogue block.
* 
Screen Shot 2021-02-10 at 3 39 39 PM

* From here, you can add and remove the timestamp
* Confirm it gets the timestamp value from the current time in the player. If the player is in the `10:00` position, the timestamp should get this value by clicking on the `Add timestamp` link.
* Open the block settings.
* You should see the Range control. Confirm you can change the timestamp using it.
* Confirm that when clicking on the timestamp button, it playback the player to its value. In the following screenshot, the player should playback at `14:16`.
*
image

#### Proposed changelog entry for your changes:

*

### CODE REVIEW

**Comment by sfougnier** [comment]
Noticed a few things:
- When I don't have an audio source in the canvas, the ability to add a timestamp is odd. Can the timestamp functionality be removed when there's no audio source to tie it to?
- A text link instead of a button feels a bit more refined and in-line with other blocks.
- We should remove the toggle from the block sidebar and just showing the disabled control since it's technically not a toggle anymore.
- I'm curious about the remove/edit states in the canvas, maybe we could try something like the attached video to mirror the sidebar?

Document -_ Transcription-1
Document -_ Transcription


https://user-images.githubusercontent.com/67021691/106968909-cf043400-6717-11eb-8e4e-b3a2962f56ff.mov

**Comment by retrofox** [comment]
- [ ] When I don't have an audio source in the canvas, the ability to add a timestamp is odd. Can the timestamp functionality be removed when there's no audio source to tie it to?
- [ ] A text link instead of a button feels a bit more refined and in-line with other blocks.
- [ ] We should remove the toggle from the block sidebar and just showing the disabled control since it's technically not a toggle anymore.
- [ ] I'm curious about the remove/edit states in the canvas, maybe we could try something like the attached video to mirror the sidebar?

**Comment by retrofox** [comment]
_**A text link instead of a button feels a bit more refined and in-line with other blocks.**_

before | after
-------|-------
image | image

before | after
-------|------
image | image

**Comment by retrofox** [comment]
> I'm curious about the remove/edit states in the canvas, maybe we could try something like the attached video to mirror the sidebar?

Personally, I like this dropdown more than adding a button in the canvas because it doesn't pollute it, in the sense looks more similar to the front-end. Sticking a Timestamp button in the canvas creates an inconsistency between both sides, IMO. Also, the dropdown would provide extra space to add other components.

That being said that, we should get a final decision for this version, because this PR just does the opposite: replace the dropdown with the button in the canvas.

**Comment by retrofox** [comment]
Other comments / questions

* The  Input control provided by the Range control only supports the `Number` type, meaning it won't be possible to print there the timestamp in `HH:MM:SS` format, but in seconds:



We could try to create something similar but it should be from scratch. I presume there isn't enough space to add three Number controls just next to the Range control, so we should add a _**simple text input**_ element, and control its content to allow only `HH:MM:SS` format. Something that I think gets tricky for v1.

* There is no way for the user to check if the transcript text is properly synced with the audio. This could get frustrating. What do you think to print the timestamp with a button that will playback at the timestamp? It will be also consistent with the front-end.

**Comment by retrofox** [comment]
> > I'm curious about the remove/edit states in the canvas, maybe we could try something like the attached video to mirror the sidebar?
> 
> Personally, I like this dropdown more than adding a button in the canvas because it doesn't pollute it, in the sense looks more similar to the front-end. Sticking a Timestamp button in the canvas creates an inconsistency between both sides, IMO. Also, the dropdown would provide extra space to add other components.

Oh, I got confused. We still need the initial `Add timestamp` button.

**Comment by retrofox** [comment]
> There is no way for the user to check if the transcript text is properly synced with the audio. This could get frustrating. What do you think to print the timestamp with a button that will playback at the timestamp? It will be also consistent with the front-end.

I took the freedom the try it.

https://user-images.githubusercontent.com/77539/107031038-57beb680-6790-11eb-80ae-de3f8b4edb6c.mov

It's already on this branch, but we can discard it in case we don't want it.

**Comment by kwight** [comment]
@sfougnier do you have any final comments here? We should get this in asap otherwise.

**Comment by sfougnier** [comment]
Looks good for our v1. LGTM

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/timestamp-control.js [hunk 1]
```diff
@@ -54,63 +57,86 @@ function setTimestampValue( typeValue, smh ) {
...
 				min={ 0 }
-				max={ 23 }
-				onChange={ hour => (
-					! isDisabled && onChange( setTimestampValue( { hour }, smh ) )
-				) }
-				disabled={ isDisabled }
+				max={ duration }
+				onChange={ ( timeInSecons ) => {
+					setRangeValue( timeInSecons );
+					debouncedSetAttributes( timeInSecons, onChange );
```
Comment (pablinos) [inline projects/plugins/jetpack/extensions/blocks/dialogue/components/timestamp-control.js:134]: Typo with `timeInSecons`

**Review by pablinos — DISMISSED** [review]
Minor issue with a variable name, but otherwise this looks great. There are some other little issues that I saw in testing, but they're not related to this change, so I think this is good to go.

File: projects/plugins/jetpack/extensions/blocks/dialogue/components/timestamp-control.js (hunk 1 above)
Comment (retrofox) [reply projects/plugins/jetpack/extensions/blocks/dialogue/components/timestamp-control.js]: addressed here https://github.com/Automattic/jetpack/pull/18717/commits/246d1aaf7677758d374146b82442fd0709790e60

**Review by allilevine — APPROVED** [review]
Works as expected:
- The "add timestamp" button appears when I focus a Dialogue block, and disappears if I don't add a timestamp and focus away.
- I can edit the timestamp in the Block settings, using fields or the range control.
- I can remove timestamps using the "remove" button.
- Clicking "add timestamp" adds the current time from the player.
- Clicking on a timestamp starts the player again.
- The timestamps display correctly on the post, and clicking them goes to that time in the player.

**Review by jeherve — COMMENTED** [review]
I can't seem to find any timestamp controls at all, I must be missing something.

image

Am I missing a specific part of the instructions?

**Comment by retrofox** [comment]
> I can't seem to find any timestamp controls at all, I must be missing something.
> 
> image
> 
> Am I missing a specific part of the instructions?

fair enough. Let me update the testing instructions. You are not the first with this problem. My bad.

**Review by jeherve — APPROVED** [review]
This seems to work for me, but I do have a few questions about the UX. Those are just my personal experiences with the block though, and I'm biased since I've seen previous variations of that block. Still, I thought I would share.

I'll approve anyway since it does work as expected.

********

Does this mean that from now on, you won't be able to use timestamps without an accompanying podcast player?

********

After using the forward/backwards buttons to get to the content I'm interested in, I would expect a button allowing me to set that dialogue to use the timestamp I found:

image

Instead, I have to enter the timestamp manually myself.

********

After using the forward/backwards buttons to get to the content I'm interested in, I clicked on the timestamp, expecting to be able to set it there, but instead it reset the podcast player to 00:00


https://user-images.githubusercontent.com/426388/107560549-c3b07d00-6bdd-11eb-91fb-880c2e2ac00a.mov

Only after running into that problem a few times did I realize that the block sidebar was the only place where I could set the timestamp. It was a bit confusing and frustrating.

Edit: I now realize I can also click "remove", and then "add timestamp" to get the timestamp at the right timing.

********

I found the rangecontrol a bit confusing at first, I don't think I would expect that kind of input when there is already a more precise input field right above:

image

**Comment by retrofox** [comment]
> This seems to work for me, but I do have a few questions about the UX. Those are just my personal experiences with the block though, and I'm biased since I've seen previous variations of that block. Still, I thought I would share.

🍿 

> I'll approve anyway since it does work as expected.

Thanks for that 👍 

> Does this mean that from now on, you won't be able to use timestamps without an accompanying podcast player?

Correct, you won't. It was discussed and designer folks decided to do it this way.

> After using the forward/backwards buttons to get to the content I'm interested in, I would expect a button allowing me to set that dialogue to use the timestamp I found:

> Instead, I have to enter the timestamp manually myself.

Long spoken too. We suggested different approaches to avoid typing the timestamp manually. Considering that transcription could get hundreds of blocks, this labor could get annoying.
Keep in-sync or set at the current-time of the audio player would help.
However, it worths it mentioning in some way is doable in this implementation. You need to remove and add the timestamp since it gets the value of the current-time of the player when it's added.


> After using the forward/backwards buttons to get to the content I'm interested in, I clicked on the timestamp, expecting to be able to set it there, but instead it reset the podcast player to 00:00

Technically, it playbacks in the timestamp position. In your example, it sets the player at the `00:00` position.

> Only after running into that problem a few times did I realize that the block sidebar was the only place where I could set the timestamp. It was a bit confusing and frustrating.

Agree. Let's try to improve it.

> 
> Edit: I now realize I can also click "remove", and then "add timestamp" to get the timestamp at the right timing.

✋ 

> 
> I found the rangecontrol a bit confusing at first, I don't think I would expect that kind of input when there is already a more precise input field right above:

The idea is to replace the `HH:MM:SS` controls with another one, in follow-up tasks.

**Comment by retrofox** [comment]
r220868-wpcom

**Comment by retrofox** [comment]
> After using the forward/backwards buttons to get to the content I'm interested in, I would expect a button allowing me to set that dialogue to use the timestamp I found:

https://github.com/Automattic/jetpack/pull/18785

**Comment by jeherve** [comment]
Cherry-picked to `jetpack/branch-9.4` in 2787fa155d2d319f1b85dd4c38305851d9704aa6

**Pull request merged**